core/lib/vm_interface/src/vm.rs

```diff
@@ -75,7 +77,12 @@ impl<T: VmInterface> VmInterfaceExt for T {}
 /// Encapsulates creating VM instance based on the provided environment.
 pub trait VmFactory<S>: VmInterface {
     /// Creates a new VM instance.
-    fn new(batch_env: L1BatchEnv, system_env: SystemEnv, storage: StoragePtr<S>) -> Self;
+    fn new(
+        batch_env: L1BatchEnv,
+        system_env: SystemEnv,
+        storage: StoragePtr<S>,
+        pubdata_builder: Option<Rc<dyn PubdataBuilder>>,
```

Comment:
- Dumb question: Why is `pubdata_builder` optional? Isn't publishing pubdata a mandatory part of finishing a batch? (Initially, I thought that it could be set to `None` for oneshot executor, but it's set there as well.)
- Related: could it be possible to move the `PubdataBuilder` arg to `VmInterface::finish_batch`, so that its lifecycle is clearer? (And maybe it could be possible to supply it by ref in this case, although not 100% sure it'll work for old VMs.) Or can pubdata be requested at any point?

Reply:
it's optional for older versions, they don't require it. I will investigate possibility of moving it to `finish_batch`, from first glance it can be done

Reply:
I have moved it to `finish_batch`